include/envoy/router/router.h

```diff
@@ -133,7 +133,13 @@ class CorsPolicy {
   /**
    * @return bool Whether CORS is enabled for the route or virtual host.
    */
-  virtual bool enabled() const PURE;
+  virtual const envoy::api::v2::core::RuntimeFractionalPercent& enabled() const PURE;
```

Comment:
Can this still return a bool? Same below? I think the runtime part can also be an impl detail? If not, the function doc above needs to change.

Reply:
Testing out if we can maintain the existing return type here. I think it's possible, my only concern is the use of runtime in the router config implementation. I'll test out the sandbox and update if it works :)

Reply:
It looks like if I change the return type back to a boolean and move the runtime logic into the router config implementation the runtime changes won't be picked up. I think we need to change the return type here but I could be missing something more obvious.

What do you think @mattklein123 ?

Reply:
I'm not sure I follow. If you look at config_impl.cc there are a bunch of examples of using runtime from within the route config impl?

Reply:
It's possible to use runtime in the route config implementation but, from my understanding, the runtime values will only be evaluated once at initialization. This means that if the runtime values change the route config implementation does not reload and when the cors filter is applied the enabled values are still set to the original ones.
e.g.
```bash
ront-envoy_1_87ead4946641 | [2018-12-29 05:19:44.189][000009][info][main] [source/server/server.cc:229]   transport_sockets.upstream: envoy.transport_sockets.alts,envoy.transport_sockets.capture,raw_buffer,tls
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.198][000009][info][main] [source/server/server.cc:271] admin address: 0.0.0.0:8001
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.199][000009][info][config] [source/server/configuration_impl.cc:50] loading 0 static secret(s)
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.199][000009][info][config] [source/server/configuration_impl.cc:56] loading 1 cluster(s)
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.200][000009][info][config] [source/server/configuration_impl.cc:67] loading 1 listener(s)
------
front-envoy_1_87ead4946641 | Enabled: 0
front-envoy_1_87ead4946641 | Shadow Enabled: 0
------
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.204][000009][info][config] [source/server/configuration_impl.cc:92] loading tracing configuration
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.205][000009][info][config] [source/server/configuration_impl.cc:112] loading stats sink configuration
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.205][000009][info][main] [source/server/server.cc:463] starting main dispatch loop
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.268][000009][info][upstream] [source/common/upstream/cluster_manager_impl.cc:136] cm init: all clusters initialized
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.269][000009][info][main] [source/server/server.cc:435] all clusters initialized. initializing init manager
front-envoy_1_87ead4946641 | [2018-12-29 05:19:44.269][000009][info][config] [source/server/listener_manager_impl.cc:961] all dependencies initialized. starting workers
```

Reply:
Have a look at `loader_` in config_impl.cc as well as `evaluateRuntimeMatch`. You will see that we already dynamically consult runtime during route decisions. I'm 99% sure as long as you plumb runtime into the cors policy, you can do the same thing there, and then hide all of the runtime stuff within the implementation which IMO is better.

Reply:
I definitely agree about hiding runtime stuff in the implementation. I pushed [a patch](https://github.com/envoyproxy/envoy/pull/5265/commits/170588b4c58fbb94520ffccab8437c9a30055ba6) but when I tested this against the cors sandbox I couldn't get runtime changes to be picked up. Maybe I'm missing something here.